### ISSUE TITLE
Profiling is hard for SIMD

### ISSUE THREAD

**Issue by GunnarFarneback** [eschnett/SIMD.jl#13]
I'm having trouble getting useful profiling information from functions using SIMD. No idea whether this is a shortcoming of the SIMD package or something else in Julia, but I'll start here.
Consider this simple (and nonsensical) function:
```julia
using SIMD

function f(x)
    s = Vec{4, Float32}(0)
    for i = 1:10000
        for k = i:4:(length(x)-3)
            s += vload(Vec{4, Float32}, x, k)
        end
    end
    return s
end
```
Loading it from file and doing a profiling gives
```julia
julia> include("/tmp/f.jl")
f (generic function with 1 method)                                                         
                                                                                           
julia> x=rand(Float32, 1000000);
                                                                                           
julia> f(x);
                                                                                           
julia> @profile f(x);
                                                                                           
julia> Profile.print(format=:flat)
 Count File                            Line Function                                       
  2778 ./<missing>                       -1 anonymous                                      
  2778 ./REPL.jl                         62 eval_user_input(::Any, ::Base.REPL.REPLBac...  
  2778 ./REPL.jl                         92 macro expansion                                
  2778 ./boot.jl                        234 eval(::Module, ::Any)                          
  2778 ./event.jl                        46 (::Base.REPL.##1#2{Base.REPL.REPLBackend})()   
  2778 ./profile.jl                      16 macro expansion;                               
  1928 ...julia/v0.5/SIMD/src/SIMD.jl   993 +                                              
   849 ...julia/v0.5/SIMD/src/SIMD.jl     6 Type                                           
  1928 ...julia/v0.5/SIMD/src/SIMD.jl     7 Type                                           
  3856 ...julia/v0.5/SIMD/src/SIMD.jl   548 llvmwrap                                       
     4 ...julia/v0.5/SIMD/src/SIMD.jl     6 macro expansion                                
  1924 ...julia/v0.5/SIMD/src/SIMD.jl   567 macro expansion                                
  1928 ...julia/v0.5/SIMD/src/SIMD.jl  1261 vload                                          
  2777 /tmp/f.jl                          4 f(::Array{Float32,1})                          
```
All time gets pinned on the first line using any SIMD instruction (line 4), which is obviously bogus in this example and in general makes the profiling information useless.

**Comment by eschnett** [comment]
On a side note: I looked at the generated assembler code (x64_86, LLVM 3.8.0), and handling the loop bounds of the inner loop is not well optimized. This is the inner loop:
```
Source line: 567
L144:
	vaddps	(%rdx), %xmm0, %xmm0
Source line: 4
	testq	%rax, %rax
	sete	%bl
	orb	%cl, %bl
	movzbl	%bl, %esi
	addq	$16, %rdx
	addq	$-4, %rax
	cmpl	$1, %esi
	jne	L144
```
The four instructions `sete`, `orb`, `movzbl`, and `cmpl` don't need be there.

However, this is unrelated to SIMD -- this has to do with how stepranges are implemented in Julia, and LLVM missing certain optimizations.

**Comment by eschnett** [comment]
Regarding the profile information: SIMD does not touch line number information in any way. The generated assembler code shows line numbers mostly for `SIMD.jl`, and not for the loop in `f.jl`. I assume that the profiler ultimately looks at the same line number information.

There have been a few discussions regarding line numbers and inlined code. (SIMD uses a lot of inlining.) Maybe this is a generic problem, caused by the recent changes? You might want to read through some of these email threads and/or attach your findings there.

Apart from confirming the issue and these generic statements I'm out of my depth and can't help you further.

**Comment by GunnarFarneback** [comment]
On 0.6 this now gives
```
 Count File                            Line Function                                      
  2390 ./<missing>                       -1 anonymous                                     
  2390 ./REPL.jl                         66 eval_user_input(::Any, ::Base.REPL.REPLBac... 
  2390 ./REPL.jl                         97 macro expansion                               
  2390 ./boot.jl                        234 eval(::Module, ::Any)                         
  2390 ./event.jl                        68 (::Base.REPL.##3#4{Base.REPL.REPLBackend})()  
  2390 ./loading.jl                     481 include_from_node1(::String)                  
  2390 ./profile.jl                      16 macro expansion;                              
     1 ./range.jl                        18 Type                                          
     1 ./range.jl                        70 Type                                          
     1 ./range.jl                       115 colon                                         
     1 ./range.jl                        37 steprange_last(::Int64, ::Int64, ::Int64)     
  1676 ...julia/v0.6/SIMD/src/SIMD.jl   993 +                                             
  3352 ...julia/v0.6/SIMD/src/SIMD.jl   548 llvmwrap                                      
  1673 ...julia/v0.6/SIMD/src/SIMD.jl   567 macro expansion                               
   714 /tmp/f.jl                          6 f(::Array{Float32,1})                         
  1676 /tmp/f.jl                          7 f(::Array{Float32,1})                         
     3 /tmp/f.jl                          6 macro expansion                               
```
which makes a lot more sense. (I don't get the last line about a macro expansion on line 6, but seems in any case minor compared to the original problem.)

Fixed by JuliaLang/julia#18180. Thanks @yuyichao.
